### PR TITLE
ENH: Add StringMethod.find and rfind

### PR DESCRIPTION
**Pull request by sinhrks**
Derived from #9111.

### CODE REVIEW

File: pandas/core/strings.py [hunk 1]
```diff
@@ -544,6 +544,46 @@ def str_findall(arr, pat, flags=0):
...
+    Parameters
+    ----------
+    sub : str
+        Substring being searched
+    start : int
+        Left edge index
+    end : int
+        Right edge index
+    side : {'left', 'right'}, default 'left'
+
```
Comment (shoyer) [inline pandas/core/strings.py]: I would document this as switching between `find` (left side) and `rfind` (right side).

File: pandas/core/strings.py [hunk 2]
```diff
@@ -544,6 +544,46 @@ def str_findall(arr, pat, flags=0):
...
+    found : array
+    """
+
+    if not isinstance(sub, compat.string_types):
+        msg = 'substring must be a str, not {0}'
+        raise TypeError(msg.format(type(sub).__name__))
+
+    if side == 'left':
+        if end is None:
+            f = lambda x: x.find(sub, start)
```
Comment (shoyer) [inline pandas/core/strings.py]: To make the structure clearer here/less redundant, maybe try:

```
if side == 'left':
    method = 'find'
elif side == 'right':
    method = 'rfind'
else:
    raise ...

if end is None:
    f = lambda x: getattr(x, method)(sub, start)
else:
    ...
```

**Comment by sinhrks** [comment]
Thanks, fixed.

**Comment by shoyer** [comment]
OK, looks good to me.

File: pandas/core/strings.py [hunk 3]
```diff
@@ -544,6 +544,46 @@ def str_findall(arr, pat, flags=0):
     return _na_map(regex.findall, arr)
 
 
+def str_find(arr, sub, start=0, end=None, side='left'):
+    """
+    Return indexes in each strings where the substring is
+    wholly contained between [start:end]. Return -1 on failure.
```
Comment (jorisvandenbossche) [inline pandas/core/strings.py]: 'wholly' sounds a bit odd to me (non-native speaker), is 'fully' better?

File: pandas/core/strings.py [hunk 4]
```diff
@@ -1055,6 +1095,34 @@ def get_dummies(self, sep='|'):
...
+    sub : str
+        Substring being searched
+    start : int
+        Left edge index
+    end : int
+        Right edge index
+
+    Returns
+    -------
+    found : array
```
Comment (jorisvandenbossche) [inline pandas/core/strings.py]: Can you add a "See also" here pointing to the other function (find <-> rfind)?

**Comment by jorisvandenbossche** [comment]
@sinhrks I added some documentation comments. Further:

- I think you may group all the StringMethods enhancement in a section under 'new features' (as it starts to become a large list)

- Maybe we should start adding `versionadded` directives (see [here](http://sphinx-doc.org/markup/para.html#directive-versionadded)) to newly added functions? We don't do this consistently right now, but I think we should do that more.

**Comment by jorisvandenbossche** [comment]
and another:

- maybe add a 'Notes' section saying it is equivalent to the standard library's `str.find` (I even think you can link to that using `` :func:`str.find` ``

**Comment by sinhrks** [comment]
Modified once.

- About release note, let me do once finish #9111 (added as AI).
- Adding `Notes` is duplicated with the "Method Summary" in `text.rst`. How about adding `Notes` in each docstring, and change "Method Summary" section to a simple link to "String handling" section in `api.rst`.

File: pandas/core/strings.py [hunk 5]
```diff
@@ -1055,6 +1095,40 @@ def get_dummies(self, sep='|'):
...
+
+    See Also
+    --------
+    %(also)s
+    """)
+
+    @Appender(_shared_docs['find'] % dict(side='lowest',
+              also='rfind : Return highest indexes in each strings'))
```
Comment (jorisvandenbossche) [inline pandas/core/strings.py]: I think `rfind` has to be `Series.str.rfind` (and same below)

**Comment by jorisvandenbossche** [comment]
And on the note, I don't think that is a problem that is would be duplicated (that is the point of having both docstrings and narrative docs, there will always be some duplication). 

If the table in text.rst can be replaced by an autosummary table is separate issue I think. Positive side is that it would add the explanation manually, but at the moment I like more how the table looks in text.rst (better overview) than how it is on the api pages.

**Comment by jorisvandenbossche** [comment]
It is maybe not necessary to add it in a Note, just adding 'equivalent to standard library's str.find' in the explanation in the beginning of the docstring is also good.

**Comment by sinhrks** [comment]
Maybe my word choice is inappropriate. What I'm caring is uniformity of styles and descriptions. What I expect is all the docstring should have similar format, and sufficient information in a single place.

In this respect, current doc should be revised in separate PR to cover:

- All the docstring should provide "better overview" and this should be used both in "text.rst" and "api.rst".
- Add descriptions for standard ``string`` methods (not only `find` and `rfind`). 

This is one thing, and added a description for this PR.

**Comment by jreback** [comment]
thanks @sinhrks

**Pull request merged**